File: scripts/generate-models/transform_async_to_openapi.py
import yaml
import re
import os 

dir_path = os.path.dirname(os.path.realpath(__file__))

# Add RT prefix to some names to dedupe from batch
map_names = [
    "TranscriptionConfig",
    "TranslationConfig",
    "SpeakerChangeSensitivity",
    "DiarizationConfig",
    "SpeakerDiarizationConfig",
    "RecognitionResult",
    "RecognitionMetadata",
    "RecognitionAlternative",
    "RecognitionDisplay"
]

# Open the async spec
with open(f"{dir_path}/../../schemas/realtime.yml", 'r') as stream:
    try:
        spec = stream.read()
        for item in map_names:
            # We can't do a naive replace as DiarizationConfig is a substring of SpeakerDiarizationConfig
            spec = re.sub(f"(?![A-Za-z])(.)({item})", r"\1Realtime\2", spec)
        async_spec = yaml.safe_load(spec)
    except yaml.YAMLError as exc:
        print(exc)

# Open a basic openapi template as starting document
with open(f"{dir_path}/../../schemas/batch.yml", 'r') as stream:
    try:
        spec = stream.read()
        for item in map_names:
            # We can't do a naive replace as DiarizationConfig is a substring of SpeakerDiarizationConfig
            spec = re.sub(f"(?![A-Za-z])(.)({item})", r"\1Batch\2", spec)
        batch_spec = yaml.safe_load(spec)
    except yaml.YAMLError as exc:
        print(exc)

# Open a basic openapi template as starting document
with open(f"{dir_path}/template-openapi.yaml", 'r') as stream:
    try:
        template = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

messages_models_yaml = async_spec['components']['messages']

# We generate an enum of all message types as it is useful to have them all in one place
message_enum = []

# Add the payload field of 'components.messages' as schemas to the generated openapi spec
template["components"] = {"schemas": {}}
for model_name, model_content in messages_models_yaml.items():
    message_enum.append(model_name)
    payload = model_content['payload']
    template['components']['schemas'][model_name] = payload

# Append enum to the schemas
template['components']['schemas']['RealtimeMessage'] = {
    'type': "object",
    'properties': {
        'message': {
            'type': "string",
            'enum': message_enum
        }
    },
}

# Add the schemas from async spec to the openapi generated spec
template['components']['schemas'].update(async_spec['components']['schemas'])
template['components']['schemas'].update(batch_spec['definitions'])

# Save the generated openapi spec
with open(f"{dir_path}/openapi-transformed.yaml", 'w') as outfile:
    yaml.dump(template, outfile)

File: packages/real-time-client/scripts/transform-async-to-openapi.ts
#!/usr/bin/env tsx

import YAML from 'yaml';
import * as fs from 'node:fs';
import * as path from 'node:path';

const dirPath = path.dirname(__filename);

try {
  // Open the async spec
  const asyncSpecPath = path.join(dirPath, '../schema/realtime.yml');
  const asyncSpecContent = fs.readFileSync(asyncSpecPath, 'utf8');
  const asyncSpec = YAML.parse(asyncSpecContent);

  // Open a basic OpenAPI template as a starting document
  const templatePath = path.join(dirPath, '../schema/template-openapi.yml');
  const templateContent = fs.readFileSync(templatePath, 'utf8');
  const template = YAML.parse(templateContent);

  // Extract the messages models from async spec
  const messagesModelsYaml = asyncSpec.components.messages;

  // Generate an enum of all message types
  const messageEnum: string[] = [];

  // Initialize the schemas component if not already initialized
  if (!template.components) {
    template.components = { schemas: {} };
  } else if (!template.components.schemas) {
    template.components.schemas = {};
  }

  // Add the payload field of 'components.messages' as schemas to the generated OpenAPI spec
  for (const [modelName, modelContent] of Object.entries(messagesModelsYaml)) {
    messageEnum.push(modelName);
    if (
      modelContent &&
      typeof modelContent === 'object' &&
      'payload' in modelContent
    ) {
      const payload = modelContent.payload;
      template.components.schemas[modelName] = payload;
    } else {
      throw new Error(`Unexpected content: ${JSON.stringify(modelContent)}`);
    }
  }

  // Append enum to the schemas
  template.components.schemas.RealtimeMessage = {
    type: 'object',
    properties: {
      message: {
        type: 'string',
        enum: messageEnum,
      },
    },
  };

  // Add the schemas from async spec to the OpenAPI generated spec
  Object.assign(template.components.schemas, asyncSpec.components.schemas);

  // Save the generated OpenAPI spec
  const outputFilePath = path.join(dirPath, 'openapi-transformed.yaml');
  fs.writeFileSync(outputFilePath, YAML.stringify(template), 'utf8');
} catch (err) {
  console.error(err);
}
